src/core/lib/slice/weak_slice_hash_table.h

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2016 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef GRPC_CORE_LIB_SLICE_WEAK_SLICE_HASH_TABLE_H
+#define GRPC_CORE_LIB_SLICE_WEAK_SLICE_HASH_TABLE_H
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/lib/gprpp/memory.h"
+#include "src/core/lib/gprpp/ref_counted.h"
+#include "src/core/lib/gprpp/ref_counted_ptr.h"
+#include "src/core/lib/slice/slice_internal.h"
+
+/// Weak hash table implementation.
+///
+/// This entries in this table are weak: an entry may be removed at any time due
+/// to a number of reasons: memory pressure, hash collisions, etc.
+///
+/// The keys are \a grpc_slice objects. The values are of arbitrary type.
+
+namespace grpc_core {
+
+template <typename T>
+class WeakSliceHashTable : public RefCounted<WeakSliceHashTable<T>> {
+ public:
+  struct Entry {
+    grpc_slice key;
+    T value;  // Must not be null.
```

Comment:
I'm wondering why there is this restriction. What if you wanted to indicate something about this particular key by nullptr or the integer value 0?